volumes.cpp

```diff
@@ -1301,18 +1321,156 @@ SurfaceVolume& makeSurface( const SurfaceCard* card, VolumeCache* v){
+  if( mnemonic == "rcc" ){
+    return rccFacet( args, facet );
+  }
+  else if(mnemonic == "box"){
+    return boxFacet( args, facet );
+  }
+  else if( mnemonic == "rpp"){
+    return rppFacet( args, facet );
+  }
+  else if( mnemonic == "hex" || mnemonic == "rhp" ){
+    return hexFacet( args, facet );
+  }
+  else if( mnemonic == "rec" ){
+    return recFacet( args, facet );
+  }
+  throw std::runtime_error( mnemonic + " does not have macrobody facet support at this time." );
+}    
+
+
+//This function is used to find D in the planar equation Ax + By + Cz - D = 0 
+//when given two vectors.
+//The positioning vector always uses the first three arguments; i determines 
+//the first argument of the vector to which the plane will be parallel.  If 
+//end is true, the plane is at the end of the second vector in relation to the 
+//end of the first vector, otherwise it intersects the end of the first vector.
+double planePoint( std::vector< double > args, int i, bool end ){
+  if( end ){
+    return args.at(i)*( args.at(0) + args.at(i) ) + args.at(i + 1)*( args.at(1) + args.at(i + 1) ) + args.at(i + 2)*( args.at(2) + args.at(i + 2) );
+  }
+  else{
+    return args.at(i)*args.at(0) + args.at(i + 1)*args.at(1) + args.at(i + 2)*args.at(2);
+  }
+}
+
+
+SurfaceVolume* rccFacet( const std::vector< double > args, int facet ){
+  if( facet == 1 ){
+    //cylinder surface
+    return new RccVolume( Vector3d(args), Vector3d(args,3), args.at(6), true );
+  }
+  else if( facet == 2 || facet == 3 ){
+    //the two end planes
+    bool end = ( facet == 2);
+    Vector3d v( args.at(3), args.at(4), args.at(5) );
+    double D = planePoint( args, 3, end );
+    return new PlaneSurface( v, D/v.length() );
+  }
+  else{
+    throw std::runtime_error( "rcc only has 3 facets" );
+  }
+}
+
+SurfaceVolume* boxFacet( const std::vector< double > args, int facet ){
+  //This function creates the planes for the facets of box.
+  if( facet >= 1 && facet <= 6 ){
+    //if the facet number is odd, the plane is at the end of the vector specified
+    //which consists of arguments 3,4,5 for 1 & 2, 6,7,8 for 3 & 4, and so on.
+    int  i = ( ( facet + facet%2 ) / 2 ) * 3;
+    bool  end = ( facet%2 == 1 );
+    Vector3d v( args.at(i), args.at(i + 1), args.at(i + 2) );
+    double D = planePoint( args, i, end );
+    return new PlaneSurface( v, D/v.length() );
+  }
+  else{
+    throw std::runtime_error( "box only has 6 facets");
+  }
 }
 
+SurfaceVolume* rppFacet( const std::vector< double > args, int facet ){
+  if( facet >= 1 || facet <= 6 ){
+  //The vector3d determines which plane; 1 and 2 becomes (1,0,0), 2 and 3 (0,1,0), etc
+    Vector3d v( 0, 0, 0 );
+    v.v[facet/2 - 1 + facet%2] = 1;
+    //The assignment of the facets is kind of backwards given the ordering of the arguments,
+    //so there needs to be interesting math to get 1 0 3 2 5 4.
+    return new PlaneSurface( v, args.at( facet - 2 + 2*( facet%2 ) ) );
```

Comment:
I think D = facet/2 + facet%2 assuming integer math.  It could be more clear, perhaps, if you introduced
`direction = facet/2` above and then used
`direction + facet%2` as the argument here